helix-term/src/handlers/signature_help.rs

```diff
@@ -0,0 +1,345 @@
+        match event {
+            SignatureHelpEvent::Invoked => {
+                self.trigger = Some(SignatureHelpInvoked::Manual);
+                self.state = State::Closed;
+                self.finish_debounce();
+                None
+            }
+            SignatureHelpEvent::Trigger => {
+                if self.trigger.is_none() {
+                    self.trigger = Some(SignatureHelpInvoked::Automatic)
+                }
+                Some(Instant::now() + Duration::from_millis(TIMEOUT))
+            }
+            SignatureHelpEvent::ReTrigger => {
+                if matches!(self.state, State::Closed) {
+                    timeout
+                } else {
+                    if self.trigger.is_none() {
+                        self.trigger = Some(SignatureHelpInvoked::Automatic)
+                    }
+                    Some(Instant::now() + Duration::from_millis(TIMEOUT))
+                }
+            }
+            SignatureHelpEvent::Cancel => {
+                if !matches!(self.state, State::Closed) {
+                    self.state = State::Closed;
+                }
+                None
+            }
+            SignatureHelpEvent::RequestComplete { open: opened } => {
+                if opened {
+                    // don't cancel rerequest that were already triggered
+                    if let State::Pending { request } = &self.state {
+                        if !request.is_closed() {
+                            return timeout;
+                        }
+                    }
+                    self.state = State::Open;
+                } else {
+                    self.state = State::Closed;
+                }
+                timeout
+            }
+        }
+    }
+
+    fn finish_debounce(&mut self) {
+        let invocation = self.trigger.take().unwrap();
+        let (tx, rx) = cancelation();
+        self.state = State::Pending { request: tx };
+        job::dispatch_blocking(move |editor, _| request_signature_help(editor, invocation, rx))
+    }
+}
+
+pub fn request_signature_help(
+    editor: &mut Editor,
+    invoked: SignatureHelpInvoked,
+    cancel: CancelRx,
+) {
+    let (view, doc) = current!(editor);
+
+    // TODO merge multiple language server signature help into one instead of just taking the first language server that supports it
+    let future = doc
+        .language_servers_with_feature(LanguageServerFeature::SignatureHelp)
+        .find_map(|language_server| {
+            let pos = doc.position(view.id, language_server.offset_encoding());
+            language_server.text_document_signature_help(doc.identifier(), pos, None)
+        });
+
+    let Some(future) = future else {
+        // Do not show the message if signature help was invoked
+        // automatically on backspace, trigger characters, etc.
+        if invoked == SignatureHelpInvoked::Manual {
+            editor
+                .set_error("No configured language server supports signature-help");
```

Comment:
In general the status messages in helix-event endup doing the same thing (being put in a queue and then we set the error in main). In the future we may be able to handle that better (to avoid overlapping error for example) but for now all that would offer is just an additional roundtrip trough the channel.

This code was mostly just moved so I tried to not change it where it wasn't necessary.